protocol/superchainerc20/L1-native-factory.md

```diff
@@ -0,0 +1,307 @@
+
+## Proposed Solution
+
+In what follows, we will list each requirement and detail how the factory should look to accomplish them.
+
+### Same address: Predeploys + `CREATE3`
+
+**Predeploys**
+
+Factories will live in each L2 as a predeploy on the same address, which is crucial to achieving the same address for `SuperchainERC20`s across chains.
+
+If the same address for the `SuperchainERC20` on L1 is also needed, then the address at which the predeploy is initialized should match the factory's L1 address.
+
+**Creation method**
+
+The `SuperchainERC20` addresses should be independent of the implementations. `CREATE` was discarded due to nonce dependability, and `CREATE2` was discarded because of its dependence on the creation code. For these reasons, the factory should use `CREATE3`.
+
+The salt will be the L1 address.
+
+### Implementation Upgradability: Beacon Proxies
+
+**BeaconProxies**
+
+The factory will deploy `SuperchainERC20`s as BeaconProxies, as this is the easiest way to upgrade multiple proxies simultaneously. Each BeaconProxy will delegate-call to the implementation address provided by the Beacon Contract.
+
+**Beacon Contract**
+
+The Beacon Contract holds the implementation address for the BeaconProxies. It will be a predeploy (consistent with the current architecture of the OP stack).
+
+**Implementations**
+
+Each type of ERC20 (vanilla, vote token, rebasing, xERC20, etc) will require their unique implementation and, therefore, factory. Optimism should deploy and support the most common implementations. Anyone can deploy their custom implementations and factories but without predeploys.
+
+In an implementation update, the upgraded chains must perform a hardfork to change the constant value in the Beacon Contract representing the implementation address.
+
+### Factory Upgradability
+
+The factory will be a simple Proxy contract pointing to the current implementation. In case of an upgrade to the factory, the upgrading chains will need to deploy the new implementation and perform a hardfork to update the implementation address.
+
+### Conservation of Properties
+
+The first deployment should be initialized from a Deployer contract in L1 called `NATIVE_INITIALIZER`, with a message that includes name, symbol, and decimal information. The predeployed factory on L2 will check that the message comes from the whitelisted address of the Deployer (which will live as a state variable in the predeploy) and proceed with the rest of the `SuperchainERC20` deployment.
+
+After the first deployment, it is possible to deploy from L1 or from the factory that already deployed and stored the metadata using interop.
+
+An alternative design to deploy from L1 is to use the `L1 blockhash` along with storage proofs to prove the properties. This would allow to initialize deployments from L2 as well (the factory would need to implement two external functions in this case).
```

Comment:
Hmmm how would this work? Each ERC20 stores stuff in different locations and some ERC20s store their details in functions instead of storage variables.